core/src/main/scala/kafka/coordinator/transaction/TransactionMetadata.scala

```diff
@@ -125,50 +134,177 @@ private[coordinator] class TransactionMetadata(val pid: Long,
+  }
+
+  def prepareNewPid(updateTimestamp: Long): TransactionMetadataTransition = {
+
+    prepareTransitionTo(Empty, producerEpoch, txnTimeoutMs, immutable.Set.empty[TopicPartition], -1, updateTimestamp)
+  }
+
+  def prepareAddPartitions(addedTopicPartitions: immutable.Set[TopicPartition],
+                           updateTimestamp: Long): TransactionMetadataTransition = {
+
+    if (state == Empty || state == CompleteCommit || state == CompleteAbort) {
+      prepareTransitionTo(Ongoing, producerEpoch, txnTimeoutMs, (topicPartitions ++ addedTopicPartitions).toSet, updateTimestamp, updateTimestamp)
+    } else {
+      prepareTransitionTo(Ongoing, producerEpoch, txnTimeoutMs, (topicPartitions ++ addedTopicPartitions).toSet, txnStartTimestamp, updateTimestamp)
+    }
+  }
+
+  def prepareAbortOrCommit(newState: TransactionState,
+                           updateTimestamp: Long): TransactionMetadataTransition = {
+
+    prepareTransitionTo(newState, producerEpoch, txnTimeoutMs, topicPartitions.toSet, txnStartTimestamp, updateTimestamp)
+  }
+
+  def prepareComplete(updateTimestamp: Long): TransactionMetadataTransition = {
+    val newState = if (state == PrepareCommit) CompleteCommit else CompleteAbort
+    prepareTransitionTo(newState, producerEpoch, txnTimeoutMs, topicPartitions.toSet, txnStartTimestamp, updateTimestamp)
+  }
+
+  private def prepareTransitionTo(newState: TransactionState,
+                                  newEpoch: Short,
+                                  newTxnTimeoutMs: Int,
+                                  newTopicPartitions: immutable.Set[TopicPartition],
+                                  newTxnStartTimestamp: Long,
+                                  updateTimestamp: Long): TransactionMetadataTransition = {
     if (pendingState.isDefined)
-      throw new IllegalStateException(s"Preparing transaction state transition to $newState while it already a pending state ${pendingState.get}")
+      throw new IllegalStateException(s"Preparing transaction state transition to $newState " +
+        s"while it already a pending state ${pendingState.get}")
 
     // check that the new state transition is valid and update the pending state if necessary
     if (TransactionMetadata.validPreviousStates(newState).contains(state)) {
       pendingState = Some(newState)
-      true
+
+      TransactionMetadataTransition(producerId, newEpoch, newTxnTimeoutMs, newState, newTopicPartitions, newTxnStartTimestamp, updateTimestamp)
     } else {
-      false
+      throw new IllegalStateException(s"Preparing transaction state transition to $newState failed since the target state" +
+        s" $newState is not a valid previous state of the current state $state")
     }
   }
 
-  def completeTransitionTo(newState: TransactionState): Boolean = {
+  // TODO: instead of returning false, we could directly throw exception since it should never be expected
+  def completeTransitionTo(newMetadata: TransactionMetadataTransition): Boolean = {
```

Comment:
We discussed offline changing this to assert a valid transition instead of returning a boolean. We could do this in a follow-up if you prefer.